text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+
+Removing this mechanism will mean that instead of, theoretically, having only place consider older versions of the schema,
+we will now have multiple places in the code that should be aware of it. But thats ok, because its enforced by the compiler,
+and it also provides better context and granularity for doing that so called evolution.
+
+Also, in reality, we don't actually do any evolution. All we do is synthetically upgrade the schema version,
+without adding or modifying any properties.
+
+This brings us to the next item.
+
+##### Compatibility Validation
+
+This is done in the [`verifyManifestVersion`](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/cx-api/lib/versioning.ts#L39) function,
+and contains two validations:
+
+1. **Framework cannot be bigger than CLI.**
+
+   We have already [described](#step-3-validate-cli--framework) how to handle this validation.
+
+2. **CLI cannot be bigger than Framework.**
+
+   This validation is actually misleading, since we do support CLI > Framework. The way its currently implemented is by doing
+   the aforementioned evolution. But, if the evolution is not necessary anymore, than this validation is not necessary anymore.
+   Which makes sense, since we do support this scenario.
+
+### CLI Regression Tests
+
+The second major mechanism we need to implement is CLI regression tests.
+
+As already mentioned, `cx-protocol` API breakage is not the only thing that can go wrong.
+This is to ensure that we don't introduce breaking changes to the CLI itself. The plan here is fairly straight forward:
+
+```console
+git checkout v${LATEST_PUBLISHED_VERSION}
+npm install /path/to/aws-cdk
+npm install /path/to/packages/*
+cd packages/aws-cdk
+yarn integ-cli
+```
+
+> Not to be taken literally, commands just illustrate intent
+
+# Drawbacks
+
+The only drawback I can think of is that we will be forcing users to upgrade their CLI when they upgrade their framework.
+This is an additional action that they didn't have to take until now. However, calling this a *drawback* is debatable,
+since it actually does provide a more consistent behavior than we have now. In any case, we don't consider this a sufficiently serious issue.
+
+# Rationale and Alternatives
+
+## New Compatibility Model
+
+As we realized by now, this RFC proposes a significant change to our compatibility model.
+Instead of attempting to have new framework versions support old CLI versions,
+we will now require that CLI `>=` Framework. Forcing users to upgrade their CLI every time they upgrade their framework.
+
+### Why is it OK to enforce this requirement?
+
+- We want to migrate most of the functionality to the framework. Making CLI upgrades especially safe.
+- Users tend to upgrade the CLI before upgrading the framework.
+- We sometimes require that even now, and we haven't seen any backlash due to that.
```

Comment:
There's a difference between requiring an upgrade when there *is* new features to be gained on the way... And requiring an upgrade because two arbitrary strings no longer match.

Reply:
The thing is that for many users, the current functionality is the same as an arbitrary upgrade. A user would upgrade the framework, not use any new feature or API, or actually not even change the code at all, and still be required to upgrade because an arbitrary feature (that he/she doesn't used), was added. 

I will extend on this a bit.

Reply:
@RomainMuller seems like you are in disagreement that this proposal is the right direction to go. I am correct to understand that?

Reply:
I don't know that I'd be so categorical... But I can't help but feeling we're taking effort off the maintainers' plate and pushing it to the user's plate... I (as a maintainer) can surely appreciate the value **I** get from it, but still am bummed out by the noise to the user... We release **often**, then we "aggressively" notify users about new versions... Do we want to get into the old time's Windows dance ("Oh yeah, it's Wednesday/patch day... Gotta install patches and reboot mah machine")?

Reply:
We agreed that this is not going to block us from at least attempting to go down this road. Having said that, i am currently thinking of a mechanism to solve this issue.